# Project 1: SAT & ACT Analysis

The first markdown cell in a notebook is a great place to provide an overview of your entire project. You will likely want to at least state your

## Problem Statement

as well as an

## Executive Summary

If you want to, it's great to use relative links to direct your audience to various sections of a notebook. **HERE'S A DEMONSTRATION WITH THE CURRENT SECTION HEADERS**:

### Contents:
- [2017 Data Import & Cleaning](#Data-Import-and-Cleaning)
- [2018 Data Import and Cleaning](#2018-Data-Import-and-Cleaning)
- [Exploratory Data Analysis](#Exploratory-Data-Analysis)
- [Data Visualization](#Visualize-the-data)
- [Descriptive and Inferential Statistics](#Descriptive-and-Inferential-Statistics)
- [Outside Research](#Outside-Research)
- [Conclusions and Recommendations](#Conclusions-and-Recommendations)

**If you combine your problem statement, executive summary, data dictionary, and conclusions/recommendations, you have an amazing README.md file that quickly aligns your audience to the contents of your project.** Don't forget to cite your data sources!

*All libraries used should be added here*

In [2]:
import numpy as np
import pandas as pd
import statistics as st
import matplotlib.pyplot as plt
import math as math
import seaborn as sns
%matplotlib inline

## 2017 Data Import and Cleaning

#### 1. Read In SAT & ACT  Data

Read in the `sat_2017.csv` and `act_2017.csv` files and assign them to appropriately named pandas dataframes.

In [3]:
sat_2017 = pd.read_csv("sat_2017.csv") 
act_2017 = pd.read_csv("act_2017.csv") 

#### 2. Display Data

Print the first 10 rows of each dataframe to your jupyter notebook

In [4]:
sat_2017.head(10)

State Participation  Evidence-Based Reading and Writing  \
0               Alabama            5%                                 593   
1                Alaska           38%                                 547   
2               Arizona           30%                                 563   
3              Arkansas            3%                                 614   
4            California           53%                                 531   
5              Colorado           11%                                 606   
6           Connecticut          100%                                 530   
7              Delaware          100%                                 503   
8  District of Columbia          100%                                 482   
9               Florida           83%                                 520   

   Math  Total  
0   572   1165  
1   533   1080  
2   553   1116  
3   594   1208  
4   524   1055  
5   595   1201  
6   512   1041  
7   492    996  
8   468    950  
9   497   1017

In [5]:
act_2017.head(10)

State Participation  English  Math  Reading  Science  \
0              National           60%     20.3  20.7     21.4     21.0   
1               Alabama          100%     18.9  18.4     19.7     19.4   
2                Alaska           65%     18.7  19.8     20.4     19.9   
3               Arizona           62%     18.6  19.8     20.1     19.8   
4              Arkansas          100%     18.9  19.0     19.7     19.5   
5            California           31%     22.5  22.7     23.1     22.2   
6              Colorado          100%     20.1  20.3     21.2     20.9   
7           Connecticut           31%     25.5  24.6     25.6     24.6   
8              Delaware           18%     24.1  23.4     24.8     23.6   
9  District of Columbia           32%     24.4  23.5     24.9     23.5   

  Composite  
0        21  
1      19.2  
2      19.8  
3      19.7  
4      19.4  
5      22.8  
6      20.8  
7      25.2  
8      24.1  
9      24.2

#### 3. Verbally Describe Data

Take your time looking through the data and thoroughly describe the data in the markdown cell below. 

Answer:

#### 4a. Does the data look complete? 

The Data looks complete; however, some data anomalies are evident.

#### 4b. Are there any obvious issues with the observations?

**What is the minimum *possible* value for each test/subtest? What is the maximum *possible* value?**

Consider comparing any questionable values to the sources of your data:
- [SAT](https://blog.collegevine.com/here-are-the-average-sat-scores-by-state/)
- [ACT](https://blog.prepscholar.com/act-scores-by-state-averages-highs-and-lows)

SAT score ranges 
-	Participation% - % of total students who took the exam in the state
-	Math : expected range between minimum of 200 to maximum of 800.
-	Evidence-Based Reading and Writing (EBRW) : expected range between minimum of 200 to maximum of 800.  
-	Total Score : expected range between minimum of 400 to maximum of 1600.
-	Participation% - % of total students who took the exam in the state

ACT score ranges
-	Participation% - % of total students who took the exam in the state
-	English : expected range between minimum of 1 to maximum of 36.
-	Reading : expected range between minimum of 1 to maximum of 36.
-	Math : expected range between minimum of 1 to maximum of 36.
-	Science : expected range between minimum of 1 to maximum of 36.
-	Composite : calculated as the average of all subjects: expected range between minimum of 1 to maximum of 36.  


#### 4c. Fix any errors you identified

**The data is available** so there's no need to guess or calculate anything. If you didn't find any errors, continue to the next step.

There are some scores that did not meet the expected scores. Further research showed the correct values.
- SAT_2017
    - Maryland.Math score was 52; should be 524
- ACT_2017
    - Maryland.Science score was 2.3; should be 23.2
    - Wyoming.Composite was 20.2x; should be 20.2

#### 5. What are your data types? 
Display the data types of each feature. 

In [6]:
sat_2017.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 51 entries, 0 to 50
Data columns (total 5 columns):
State                                 51 non-null object
Participation                         51 non-null object
Evidence-Based Reading and Writing    51 non-null int64
Math                                  51 non-null int64
Total                                 51 non-null int64
dtypes: int64(3), object(2)
memory usage: 2.1+ KB


In [7]:
act_2017.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 52 entries, 0 to 51
Data columns (total 7 columns):
State            52 non-null object
Participation    52 non-null object
English          52 non-null float64
Math             52 non-null float64
Reading          52 non-null float64
Science          52 non-null float64
Composite        52 non-null object
dtypes: float64(4), object(3)
memory usage: 2.9+ KB


What did you learn?
- Do any of them seem odd?  
- Which ones are not as they should be?  

For sat_2017.csv: Column title: “Participation”;  Data Type is Object
-	This due to the “%” character in the column
-	Data Type should be  int64

For act_2017.csv: Column title: “Participation”;  Data Type is Object
-	This due to the “%” character in the column
-	Data Type should be  int64

For act_2017.csv: Column title: “Composite”;  Data Type is Object
-	This due to the value of “20.2x” for Wyoming; value should be  20.2
-	Data Type should be  int64


#### 6. Fix Incorrect Data Types
Based on what you discovered above, use appropriate methods to re-type incorrectly typed data.
- Define a function that will allow you to convert participation rates to an appropriate numeric type. Use `map` or `apply` to change these columns in each dataframe.

- Fix any individual values preventing other columns from being the appropriate type.

In [8]:
sat_2017['Participation'] = sat_2017['Participation'].map(lambda x: x.lstrip('+-').rstrip('%'))

In [9]:
act_2017['Participation'] = act_2017['Participation'].map(lambda x: x.lstrip('+-').rstrip('%'))

In [10]:
sat_2017.at[20,'Math'] = 524

In [11]:
act_2017.Composite = act_2017.Composite.map(lambda composite: composite.replace('20.2x', '20.2'))

- Finish your data modifications by making sure the columns are now typed appropriately.

In [12]:
sat_2017['Participation'] = sat_2017['Participation'].astype(int)

In [13]:
act_2017['Participation'] = act_2017['Participation'].astype(int)

In [14]:
act_2017['Composite'] = act_2017['Composite'].astype(float)

- Display the data types again to confirm they are correct.

In [15]:
sat_2017.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 51 entries, 0 to 50
Data columns (total 5 columns):
State                                 51 non-null object
Participation                         51 non-null int32
Evidence-Based Reading and Writing    51 non-null int64
Math                                  51 non-null int64
Total                                 51 non-null int64
dtypes: int32(1), int64(3), object(1)
memory usage: 1.9+ KB


In [16]:
act_2017.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 52 entries, 0 to 51
Data columns (total 7 columns):
State            52 non-null object
Participation    52 non-null int32
English          52 non-null float64
Math             52 non-null float64
Reading          52 non-null float64
Science          52 non-null float64
Composite        52 non-null float64
dtypes: float64(5), int32(1), object(1)
memory usage: 2.7+ KB


#### 7. Rename Columns
Change the names of the columns to more expressive names so that you can tell the difference the SAT columns and the ACT columns. Your solution should map all column names being changed at once (no repeated singular name-changes). **We will be combining these data with some of the data from 2018, and so you should name columns in an appropriate way**.

**Guidelines**:
- Column names should be all lowercase (you will thank yourself when you start pushing data to SQL later in the course)
- Column names should not contain spaces (underscores will suffice--this allows for using the `df.column_name` method to access columns in addition to `df['column_name']`.
- Column names should be unique and informative (the only feature that we actually share between dataframes is the state).

In [17]:
def sat17_chg_col_names(satdf):
    satdf.columns = ['state', 'sat_participation_17','sat_reading_writing_17','sat_math_17', 'sat_total_17']
    return

sat17_chg_col_names(sat_2017)

def act17_chg_col_names(actdf):
    actdf.columns = ['state', 'act_participation_17','act_english_17','act_math_17', 'act_reading_17', 'act_science_17','act_composite_17']
    return

act17_chg_col_names(act_2017)


#### 8. Create a data dictionary

Now that we've fixed our data, and given it appropriate names, let's create a [data dictionary](http://library.ucmerced.edu/node/10249). 

A data dictionary provides a quick overview of features/variables/columns, alongside data types and descriptions. The more descriptive you can be, the more useful this document is.

Example of a Fictional Data Dictionary Entry: 

|Feature|Type|Dataset|Description|
|---|---|---|---|
|**county_pop**|*integer*|2010 census|The population of the county (units in thousands, where 2.5 represents 2500 people).| 
|**per_poverty**|*float*|2010 census|The percent of the county over the age of 18 living below the 200% of official US poverty rate (units percent to two decimal places 98.10 means 98.1%)|

[Here's a quick link to a short guide for formatting markdown in Jupyter notebooks](https://jupyter-notebook.readthedocs.io/en/stable/examples/Notebook/Working%20With%20Markdown%20Cells.html).

Provided is the skeleton for formatting a markdown table, with columns headers that will help you create a data dictionary to quickly summarize your data, as well as some examples. **This would be a great thing to copy and paste into your custom README for this project.**

|Feature|Type|Dataset|Description|
|---|---|---|---|
|state|object|ACT|The name of the state|
|act_participation_17|int64|ACT|state participation (in %) in ACT for 2017|
|act_english_17|float64|ACT|english proficiency for 2017|
|act_math_17|float64|ACT|math proficiency for 2017|
|act_reading_17|float64|ACT|reading proficiency for 2017|
|act_science_17|float64|ACT|science proficiency for 2017|
|act_composite_17|float64|ACT|composition proficiency(ave. of all subjects) for 2017|
|sat_participation_17|int32|SAT| state participation (in %) in SAT for 2017|
|sat_reading_writing_17|int64|SAT|reading & writing proficiency for 2017|
|sat_math_17|int64|SAT| math proficiency for 2017|
|sat_total_17|int64|SAT|total of sat_reading_writing and sat_math for 2017|


#### 9. Drop unnecessary rows

One of our dataframes contains an extra row. Identify and remove this from the dataframe.

In [18]:
act_2017.drop([0], axis=0).head(3)

state  act_participation_17  act_english_17  act_math_17  act_reading_17  \
1  Alabama                   100            18.9         18.4            19.7   
2   Alaska                    65            18.7         19.8            20.4   
3  Arizona                    62            18.6         19.8            20.1   

   act_science_17  act_composite_17  
1            19.4              19.2  
2            19.9              19.8  
3            19.8              19.7

#### 10. Merge Dataframes

Join the 2017 ACT and SAT dataframes using the state in each dataframe as the key. Assign this to a new variable.

In [19]:
combined_2017 = pd.merge(sat_2017, act_2017, on='state', how='inner')

#### 11. Save your cleaned, merged dataframe

Use a relative path to save out your data as `combined_2017.csv`.

In [20]:
combined_2017.to_csv('..\code\combined_2017.csv', index=False)

## 2018 Data Import and Cleaning

Links to the 2018 ACT and SAT data are provided in the README. These data live in PDFs, and so you'll get to enjoy practicing some *manual* data collection. Save these data as a CSV in your `data` directory, and import, explore, and clean these data in the same way you did above. **Make sure you comment on your steps so it is clear *why* you are doing each process**.

In [21]:
sat_2018 = pd.read_csv("sat_2018.csv") 
act_2018 = pd.read_csv("act_2018.csv")

In [22]:
sat_2018.head(10)

State  Participation%  Evidence-Based Reading and Writing  \
0               Alabama               6                                 595   
1                Alaska              43                                 562   
2               Arizona              29                                 577   
3              Arkansas               5                                 592   
4            California              60                                 540   
5              Colorado             100                                 519   
6           Connecticut             100                                 535   
7              Delaware             100                                 505   
8  District of Columbia              92                                 497   
9               Florida              97                                 522   

   Math  Total  
0   571   1166  
1   544   1106  
2   572   1149  
3   576   1169  
4   536   1076  
5   506   1025  
6   519   1053  
7   492    998  
8   480    977  
9   493   1014

In [23]:
act_2018.head(10)

State  Participation%  English  Math  Reading  Science  \
0              National              55     20.2  20.5     21.3     20.7   
1               Alabama             100     18.9  18.3     19.6     19.0   
2                Alaska              33     19.8  20.6     21.6     20.7   
3               Arizona              66     18.2  19.4     19.5     19.2   
4              Arkansas             100     19.1  18.9     19.7     19.4   
5            California              27     22.5  22.5     23.0     22.1   
6              Colorado              30     23.9  23.2     24.4     23.5   
7           Connecticut              26     26.0  24.8     26.1     24.9   
8              Delaware              17     23.7  23.1     24.5     23.4   
9  District of Columbia              32     23.7  22.7     24.4     23.0   

   Composite  
0       20.8  
1       19.1  
2       20.8  
3       19.2  
4       19.4  
5       22.7  
6       23.9  
7       25.6  
8       23.2  
9       23.6

In [24]:
sat_2018.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 51 entries, 0 to 50
Data columns (total 5 columns):
State                                 51 non-null object
Participation%                        51 non-null int64
Evidence-Based Reading and Writing    51 non-null int64
Math                                  51 non-null int64
Total                                 51 non-null int64
dtypes: int64(4), object(1)
memory usage: 2.1+ KB


In [25]:
act_2018.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 52 entries, 0 to 51
Data columns (total 7 columns):
State             52 non-null object
Participation%    52 non-null int64
English           52 non-null float64
Math              52 non-null float64
Reading           52 non-null float64
Science           52 non-null float64
Composite         52 non-null float64
dtypes: float64(5), int64(1), object(1)
memory usage: 2.9+ KB


In [26]:
def sat18_chg_col_names(satdf):
    satdf.columns = ['state', 'sat_participation_18','sat_reading_writing_18','sat_math_18', 'sat_total_18']
    return

sat18_chg_col_names(sat_2018)

def act18_chg_col_names(actdf):
    actdf.columns = ['state', 'act_participation_18','act_english_18','act_math_18', 'act_reading_18', 'act_science_18','act_composite_18']
    return

act18_chg_col_names(act_2018)


In [27]:
act_2018.drop([0], axis=0).head(3)

state  act_participation_18  act_english_18  act_math_18  act_reading_18  \
1  Alabama                   100            18.9         18.3            19.6   
2   Alaska                    33            19.8         20.6            21.6   
3  Arizona                    66            18.2         19.4            19.5   

   act_science_18  act_composite_18  
1            19.0              19.1  
2            20.7              20.8  
3            19.2              19.2

In [28]:
combined_2018 = pd.merge(sat_2018, act_2018, on='state', how='inner')

In [29]:
combined_2018.to_csv('..\code\combined_2018.csv', index=False)

#### Combine your 2017 and 2018 data into a single dataframe
Joining on state names should work, assuming you formatted all your state names identically. Make sure none of your columns (other than state) have identical names. Do yourself a favor and decide if you're encoding participation rates as floats or integers and standardize this across your datasets.

Save the contents of this merged dataframe as `final.csv`.

**Use this combined dataframe for the remainder of the project**.

In [30]:
combined_2017 = pd.read_csv("combined_2017.csv") 
combined_2018 = pd.read_csv("combined_2018.csv") 

In [31]:
final = pd.merge(combined_2017, combined_2018, on='state', how='inner')

In [32]:
final

state  sat_participation_17  sat_reading_writing_17  \
0                Alabama                     5                     593   
1                 Alaska                    38                     547   
2                Arizona                    30                     563   
3               Arkansas                     3                     614   
4             California                    53                     531   
5               Colorado                    11                     606   
6            Connecticut                   100                     530   
7               Delaware                   100                     503   
8   District of Columbia                   100                     482   
9                Florida                    83                     520   
10               Georgia                    61                     535   
11                Hawaii                    55                     544   
12                 Idaho                    93                     513   
13              Illinois                     9                     559   
14               Indiana                    63                     542   
15                  Iowa                     2                     641   
16                Kansas                     4                     632   
17              Kentucky                     4                     631   
18             Louisiana                     4                     611   
19                 Maine                    95                     513   
20              Maryland                    69                     536   
21         Massachusetts                    76                     555   
22              Michigan                   100                     509   
23             Minnesota                     3                     644   
24           Mississippi                     2                     634   
25              Missouri                     3                     640   
26               Montana                    10                     605   
27              Nebraska                     3                     629   
28                Nevada                    26                     563   
29         New Hampshire                    96                     532   
30            New Jersey                    70                     530   
31            New Mexico                    11                     577   
32              New York                    67                     528   
33        North Carolina                    49                     546   
34          North Dakota                     2                     635   
35                  Ohio                    12                     578   
36              Oklahoma                     7                     530   
37                Oregon                    43                     560   
38          Pennsylvania                    65                     540   
39          Rhode Island                    71                     539   
40        South Carolina                    50                     543   
41          South Dakota                     3                     612   
42             Tennessee                     5                     623   
43                 Texas                    62                     513   
44                  Utah                     3                     624   
45               Vermont                    60                     562   
46              Virginia                    65                     561   
47            Washington                    64                     541   
48         West Virginia                    14                     558   
49             Wisconsin                     3                     642   
50               Wyoming                     3                     626   

    sat_math_17  sat_total_17  act_participation_17  act_english_17  \
0           572          1165                   100            18.9   
1           533          108

In [33]:
final.to_csv('final.csv', index=False)

## Exploratory Data Analysis


### Summary Statistics
Transpose the output of pandas `describe` method to create a quick overview of each numeric feature.

In [34]:
final.describe()

sat_participation_17  sat_reading_writing_17  sat_math_17  \
count             51.000000               51.000000    51.000000   
mean              39.803922              569.117647   556.882353   
std               35.276632               45.666901    47.121395   
min                2.000000              482.000000   468.000000   
25%                4.000000              533.500000   523.500000   
50%               38.000000              559.000000   548.000000   
75%               66.000000              613.000000   599.000000   
max              100.000000              644.000000   651.000000   

       sat_total_17  act_participation_17  act_english_17  act_math_17  \
count     51.000000             51.000000       51.000000    51.000000   
mean    1126.098039             65.254902       20.931373    21.182353   
std       92.494812             32.140842        2.353677     1.981989   
min      950.000000              8.000000       16.300000    18.000000   
25%     1055.500000             31.000000       19.000000    19.400000   
50%     1107.000000             69.000000       20.700000    20.900000   
75%     1212.000000            100.000000       23.300000    23.100000   
max     1295.000000            100.000000       25.500000    25.300000   

       act_reading_17  act_science_17  act_composite_17  sat_participation_18  \
count       51.000000       51.000000         51.000000             51.000000   
mean        22.013725       21.450980         21.519608             46.627451   
std          2.067271        1.739353          2.020695             38.014187   
min         18.100000       18.200000         17.800000              2.000000   
25%         20.450000       19.950000         19.800000              4.500000   
50%         21.800000       21.300000         21.400000             52.000000   
75%         24.150000       23.200000         23.600000             79.500000   
max         26.000000       24.900000         25.500000            100.000000   

       sat_reading_writing_18  sat_math_18  sat_total_18  \
count               51.000000    51.000000     51.000000   
mean               566.980392   557.039216   1124.019608   
std                 45.515927    49.065654     94.277779   
min                497.000000   480.000000    977.000000   
25%                534.500000   520.000000   1057.500000   
50%                552.000000   547.000000   1099.000000   
75%                616.500000   600.500000   1220.000000   
max                643.000000   655.000000   1298.000000   

       act_participation_18  act_english_18  act_math_18  act_reading_18  \
count             51.000000       51.000000    51.000000       51.000000   
mean              61.647059       20.988235    21.125490       22.015686   
std               34.080976        2.446356     2.035765        2.167245   
min                7.000000       16.600000    17.800000       18.000000   
25%               28.500000       19.100000    19.400000       20.450000   
50%               66.000000       20.200000    20.700000       21.600000   
75%              100.000000       23.700000    23.150000       24.100000   
max              100.000000       26.000000    25.200000       26.100000   

       act_science_18  act_composite_18  
count       51.000000         51.000000  
mean        21.345098         21.486275  
std          1.870114          2.106278  
min         17.900000         17.700000  
25%         19.850000         19.950000  
50%         21.100000         21.300000  
75%         23.050000         23.550000  
max         24.900000         25.600000

#### Manually calculate standard deviation

$$\sigma = \sqrt{\frac{1}{n}\sum_{i=1}^n(x_i - \mu)^2}$$

- Write a function to calculate standard deviation using the formula above

In [40]:
def mean(values):
    return sum(values)*1.0/len(values)

def stanDev(values):
    length=len(values)
    m=mean(values)
    total_sum = 0
    for i in range(length):
        total_sum += (values[i]-m)**2
    under_root = total_sum*1.0/length
    return math.sqrt(under_root)

# stan_dev = stanDev(final.sat_participation_17)
# print(stan_dev)



In [36]:
final.std(axis=0)

sat_participation_17      35.276632
sat_reading_writing_17    45.666901
sat_math_17               47.121395
sat_total_17              92.494812
act_participation_17      32.140842
act_english_17             2.353677
act_math_17                1.981989
act_reading_17             2.067271
act_science_17             1.739353
act_composite_17           2.020695
sat_participation_18      38.014187
sat_reading_writing_18    45.515927
sat_math_18               49.065654
sat_total_18              94.277779
act_participation_18      34.080976
act_english_18             2.446356
act_math_18                2.035765
act_reading_18             2.167245
act_science_18             1.870114
act_composite_18           2.106278
dtype: float64

- Use a **dictionary comprehension** to apply your standard deviation function to each numeric column in the dataframe.  **No loops**  
- Assign the output to variable `sd` as a dictionary where: 
    - Each column name is now a key 
    - That standard deviation of the column is the value 
     
*Example Output :* `{'ACT_Math': 120, 'ACT_Reading': 120, ...}`

In [45]:

sd = {i:stanDev(final[i]) for i in final.columns[1:21]}
sd

{'sat_participation_17': 34.92907076664508,
 'sat_reading_writing_17': 45.21697020437866,
 'sat_math_17': 46.65713364485503,
 'sat_total_17': 91.58351056778743,
 'act_participation_17': 31.824175751231806,
 'act_english_17': 2.3304876369363363,
 'act_math_17': 1.9624620273436781,
 'act_reading_17': 2.0469029314842646,
 'act_science_17': 1.7222161451443676,
 'act_composite_17': 2.000786081581989,
 'sat_participation_18': 37.63965409204986,
 'sat_reading_writing_18': 45.067483254615375,
 'sat_math_18': 48.582237044495976,
 'sat_total_18': 93.34891120616042,
 'act_participation_18': 33.745194881997506,
 'act_english_18': 2.4222536143202795,
 'act_math_18': 2.015707255555717,
 'act_reading_18': 2.145891884510421,
 'act_science_18': 1.8516885484833543,
 'act_composite_18': 2.0855261815801147}

In [ ]:
#Code:

Do your manually calculated standard deviations match up with the output from pandas `describe`? What about numpy's `std` method?

yes; by default, numpy uses the population STDEV./ I manually calculated using the sample STDEV.

#### Investigate trends in the data
Using sorting and/or masking (along with the `.head` method to not print our entire dataframe), consider the following questions:

- Which states have the highest and lowest participation rates for the:
    - 2017 SAT?
    - 2018 SAT?
    - 2017 ACT?
    - 2018 ACT?
- Which states have the highest and lowest mean total/composite scores for the:
    - 2017 SAT?
    - 2018 SAT?
    - 2017 ACT?
    - 2018 ACT?
- Do any states with 100% participation on a given test have a rate change year-to-year?
- Do any states show have >50% participation on *both* tests either year?

Based on what you've just observed, have you identified any states that you're especially interested in? **Make a note of these and state *why* you think they're interesting**.

**You should comment on your findings at each step in a markdown cell below your code block**. Make sure you include at least one example of sorting your dataframe by a column, and one example of using boolean filtering (i.e., masking) to select a subset of the dataframe.

Which states have the highest and lowest participation rates for the:
2017 SAT?
- highest are District of Columbia, Michigan, Connecticut, Delaware; all are 100%
- lowest is North Dakota, Mississippi, Iowa ; all are 2%

In [ ]:
final.sort_values(by=['sat_participation_17'], ascending=False, inplace=False).head(5)[['state', 'sat_participation_17']]

In [ ]:
final.sort_values(by=['sat_participation_17'], ascending=True, inplace=False).head(5)[['state', 'sat_participation_17']]

2018 SAT?
- highest are the following: all are 100%
    - Colorado, Connecticut, Delaware, Michigan, Idaho 
- lowest is North Dakota; 2%

In [ ]:
final.sort_values(by=['sat_participation_18'], ascending=False, inplace=False).head(6)[['state', 'sat_participation_18']]

In [ ]:
final.sort_values(by=['sat_participation_18'], ascending=True, inplace=False).head(2)[['state', 'sat_participation_18']]

2017 ACT?
- highest are the following: all are 100%
    - Alabama, Kentucky, Wisconsin, Utah, Tennessee, South Carolina, Oklahoma, North Carolina, Nevada, Montana
    - Mississippi, Minnesota, Louisiana, Missouri, Wyoming, Colorado, Arkansas
- lowest is Maine : 8%

In [ ]:
final.sort_values(by=['act_participation_17'], ascending=False, inplace=False).head(18)[['state', 'act_participation_17']]

In [ ]:
final.sort_values(by=['act_participation_17'], ascending=False, inplace=False).head(18)[['state', 'act_participation_17']]

In [ ]:
final.sort_values(by=['act_participation_17'], ascending=True, inplace=False).head(2)[['state', 'act_participation_17']]

2018 ACT?
- highest are the following: all are 100%
    - Alabama, Kentucky, Wisconsin, Utah, Tennessee, South Carolina, Oklahoma, Ohio, North Carolina, Nevada, Nebraska, 
    - Montana, Mississippi, Louisiana, Missouri, Wyoming, Arkansas
- lowest is Maine: 7%

In [ ]:
final.sort_values(by=['act_participation_18'], ascending=False, inplace=False).head(20)[['state', 'act_participation_18']]

In [ ]:
final.sort_values(by=['act_participation_18'], ascending=True, inplace=False).head(2)[['state', 'act_participation_18']]

Which states have the highest and lowest mean total/composite scores for the:
2017 SAT?   -  highest = Minnesota (1295) ; lowest =  District of Columbia (950)
2018 SAT?   -  highest = Minnesota (1298) ;  lowest = District of Columbia (977)
2017 ACT?   -  highest = New Hampshire (25.5) ;  lowest = Nevada (17.8)
2018 ACT?   -  highest = Connecticut (25.6) ;  lowest = Nevada (17.7)


In [ ]:
final.sort_values(by=['sat_total_17'], ascending=False, inplace=False).head(3)[['state', 'sat_total_17']]

In [ ]:
final.sort_values(by=['sat_total_17'], ascending=True, inplace=False).head(3)[['state', 'sat_total_17']]

In [ ]:
final.sort_values(by=['sat_total_18'], ascending=False, inplace=False).head(3)[['state', 'sat_total_18']]

In [ ]:
final.sort_values(by=['sat_total_18'], ascending=True, inplace=False).head(3)[['state', 'sat_total_18']]

In [ ]:
final.sort_values(by=['act_composite_17'], ascending=False, inplace=False).head(3)[['state', 'act_composite_17']]

In [ ]:
final.sort_values(by=['act_composite_17'], ascending=True, inplace=False).head(3)[['state', 'act_composite_17']]

In [ ]:
final.sort_values(by=['act_composite_18'], ascending=False, inplace=False).head(3)[['state', 'act_composite_18']]

In [ ]:
final.sort_values(by=['act_composite_18'], ascending=True, inplace=False).head(3)[['state', 'act_composite_18']]

Do any states with 100% participation on a given test have a rate change year-to-year?
- yes, Minnesota (2017 = 100%, 2018 = 99%) and Colorado (2017 = 100%, 2018 = 30%) 

In [ ]:
final.sort_values(by=['act_participation_17','act_participation_18'], ascending=False, inplace=False).head(18)[['state', 'act_participation_17', 'act_participation_18']]

Do any states show have >50% participation on both tests either year?
- yes, see, below

In [ ]:
final.query('sat_participation_17 > 50').sort_values(by=['sat_participation_17'], ascending=True, inplace=False).head(18)[['state', 'sat_participation_17', 'act_participation_17','sat_participation_18', 'act_participation_18']]

## Visualize the data

There's not a magic bullet recommendation for the right number of plots to understand a given dataset, but visualizing your data is *always* a good idea. Not only does it allow you to quickly convey your findings (even if you have a non-technical audience), it will often reveal trends in your data that escaped you when you were looking only at numbers.

Some recommendations on plotting:
- Plots have titles
- Plots have axis labels
- Plots have appropriate tick labels
- All text is legible in a plot
- Plots demonstrate meaningful and valid relationships
- Plots are interpreted to aid understanding

There is such a thing as too many plots, and there are a *lot* of bad plots. You might make some! (But hopefully not with the guided prompts below).

#### Use Seaborn's heatmap with pandas `.corr()` to visualize correlations between all numeric features

Heatmaps are generally not appropriate for presentations, and should often be excluded from reports as they can be visually overwhelming. **However**, they can be extremely useful in identify relationships of potential interest (as well as identifying potential collinearity before modeling).

*example*:
```python
sns.heatmap(df.corr())
```

Please take time to format your output, adding a title. Look through some of the additional arguments and options. (Axis labels aren't really necessary, as long as the title is informative).

In [ ]:
def getDF(data_url, columns):
    #retrieve data from url, create dataframe, return it
    data = pd.read_csv(final, names=columns)
    return data

def heatMap(df, mirror):
    
    corr = df.corr()
    fig, ax = plt.subplots(figsize=(20, 10))
    colormap = sns.diverging_palette(220, 10, as_cmap=True)
   
    if mirror == True:
        sns.heatmap(corr, cmap=colormap, annot=True, fmt=".2f")
        plt.xticks(range(len(corr.columns)), corr.columns);
        plt.yticks(range(len(corr.columns)), corr.columns)
    else:
        dropSelf = np.zeros_like(corr)
        dropSelf[np.triu_indices_from(dropSelf)] = True
        colormap = sns.diverging_palette(220, 10, as_cmap=True)
        sns.heatmap(corr, cmap=colormap, annot=True, fmt=".2f", mask=dropSelf)
        plt.xticks(range(len(corr.columns)), corr.columns);
        plt.yticks(range(len(corr.columns)), corr.columns)
   
    plt.show()
    

heatMap(final, False)
    

#### Define a custom function to subplot histograms

We have data for two tests for two years. We only have composite (and not subtest scores) for the 2018 ACT. We should write a function that will take the names of 2+ columns and subplot histograms. While you can use pandas plotting or Seaborn here, matplotlib gives you greater control over all aspects of your plots.

[Helpful Link for Plotting Multiple Figures](https://matplotlib.org/users/pyplot_tutorial.html#working-with-multiple-figures-and-axes)

Here's some starter code:

In [ ]:
def subplot_histograms(dataframe, list_of_columns, figsize=(10,10), list_of_titles=[], \
                       commonxlabel=None, commonylabel=None, list_of_xlabels=[], list_of_ylabels=[]):
    nrows = int(np.ceil(len(list_of_columns)/2))   # Makes sure you have enough rows
    fig, ax = plt.subplots(nrows=nrows, ncols=2, figsize=figsize) # You'll want to specify your figsize
    ax = ax.ravel() # Ravel turns a matrix into a vector, which is easier to iterate
    for i, column in enumerate(list_of_columns): # Gives us an index value to get into all our lists
        if len(list_of_columns) == len(list_of_titles):
                ax[i].set_title(list_of_titles[i])
        if len(list_of_columns) == len(list_of_xlabels):
                ax[i].set(xlabel=list_of_xlabels[i])
        if len(list_of_columns) == len(list_of_ylabels):
                ax[i].set(ylabel=list_of_ylabels[i])
        if commonxlabel != None:    #if commonxlabel is present, tis overwrites list_of_xlabels
                ax[i].set(xlabel=commonxlabel)
        if commonylabel != None:    #if commonylabel is present, tis overwrites list_of_ylabels               
                ax[i].set(ylabel=list_of_ylabels[i])     
                        
        ax[i].hist(dataframe[column]) # feel free to add more settings
        

#### Plot and interpret histograms 
For each of the following:
- Participation rates for SAT & ACT


In [ ]:
#subplot for 2017 Participation for SAT & ACT
subplot_histograms(final,['sat_participation_17','act_participation_17'], figsize=(15,5),commonxlabel= r'percent')

In [ ]:
#subplot for 2018 Participation for SAT & ACT
subplot_histograms(final,['sat_participation_18','act_participation_18'], figsize=(15,5),commonxlabel= r'percent')

- Math scores for SAT & ACT

In [ ]:
#subplot for SAT Math 2017 & 2018
subplot_histograms(final,['sat_math_17','sat_math_18'], figsize=(15,5),commonxlabel= r'score')

In [ ]:
#subplot for ACT Math 2017 & 2018
subplot_histograms(final,['act_math_17','act_math_18'], figsize=(15,5),commonxlabel= r'score') 

- Reading/verbal scores for SAT & ACT

In [ ]:
#subplot for SAT Evidence-Based Reading_Writing (EBRW) for 2017 & 2018
subplot_histograms(final,['sat_reading_writing_17','sat_reading_writing_18'], figsize=(15,5),commonxlabel= r'score')

In [ ]:
#subplot for ACT Reading) for 2017 & 2018
subplot_histograms(final,['act_reading_17','act_reading_18'], figsize=(15,5),commonxlabel= r'score')

#### Plot and interpret scatter plots

For each of the following:
- SAT vs. ACT math scores for 2017
- SAT vs. ACT verbal/reading scores for 2017
- SAT vs. ACT total/composite scores for 2017
- Total scores for SAT 2017 vs. 2018
- Composite scores for ACT 2017 vs. 2018

Plot the two variables against each other using matplotlib or Seaborn

Your plots should show:
- Two clearly labeled axes
- A proper title
- Using colors and symbols that are clear and unmistakable

**Feel free to write a custom function, and subplot if you'd like.** Functions save both time and space.


In [ ]:
x = final.sat_math_17
y = final.act_math_17

# Plot
plt.scatter(x, y)
plt.title('SAT vs. ACT Math scores for 2017')
plt.xlabel('SAT Math Score 2017')
plt.ylabel('ACT Math Score 2017')
plt.show()

In [ ]:
x = final.sat_reading_writing_17
y = final.act_reading_17

# Plot
plt.scatter(x, y)
plt.title('SAT vs. ACT Reading scores for 2017')
plt.xlabel('SAT Reading Score 2017')
plt.ylabel('ACT Reading Score 2017')
plt.show()

In [ ]:
x = final.sat_total_17
y = final.act_composite_17

# Plot
plt.scatter(x, y)
plt.title('SAT vs.ACT Total/COmposite scores for 2017')
plt.xlabel('SAT Total Score for 2017')
plt.ylabel('ACT Composite Score for 2017')
plt.show()

In [ ]:
x = final.sat_total_17
y = final.sat_total_18

# Plot
plt.scatter(x, y)
plt.title('SAT Total scores for 2017/2018')
plt.xlabel('SAT Total Score for 2017')
plt.ylabel('SAT Total Score for 2018')
plt.show()

In [ ]:
x = final.act_composite_17
y = final.act_composite_18

# Plot
plt.scatter(x, y)
plt.title('ACT Composite scores for 2017/2018')
plt.xlabel('ACT Composite Score for 2017')
plt.ylabel('ACT Composite Score for 2018')
plt.show()

#### Plot and interpret boxplots

For each numeric variable in the dataframe create a boxplot using Seaborn. Boxplots demonstrate central tendency and spread in variables. In a certain sense, these are somewhat redundant with histograms, but you may be better able to identify clear outliers or differences in IQR, etc.

Multiple values can be plotted to a single boxplot as long as they are of the same relative scale (meaning they have similar min/max values).

Each boxplot should:
- Only include variables of a similar scale
- Have clear labels for each variable
- Have appropriate titles and labels

In [ ]:
fig, ax=plt.subplots(figsize=(15,5))
ax.set(xlabel='SAT Reading Writing ')
ax.set(ylabel='SAT Score ')
ax.set_title('The Plot Value for SAT EBRW Reading and Writing 2017/2018')
bplot = sns.boxplot(
                 data=final[['sat_reading_writing_17','sat_reading_writing_18']], 
                 ax=ax)


In [ ]:
fig, ax=plt.subplots(figsize=(15,5))
ax.set(xlabel='SAT Math')
ax.set(ylabel='SAT Score')
ax.set_title('The Plot Value for SAT Math 2017/2018')
bplot = sns.boxplot(
                 data=final[['sat_math_17','sat_math_18']], 
                 ax=ax)

In [ ]:
fig, ax=plt.subplots(figsize=(15,5))
ax.set(xlabel='SAT Total')
ax.set(ylabel='SAT Score')
ax.set_title('The Plot Value for SAT Total 2017/2018')
bplot = sns.boxplot(
                 data=final[['sat_total_17','sat_total_18']], 
                 ax=ax)

In [ ]:
fig, ax=plt.subplots(figsize=(15,5))
ax.set(xlabel='SAT Participation')
ax.set(ylabel='Percentage')
ax.set_title('The Plot Value for SAT Participation 2017/2018')
bplot = sns.boxplot(
                 data=final[['sat_participation_17','sat_participation_18']], 
                 ax=ax)

In [ ]:
fig, ax=plt.subplots(figsize=(15,5))
ax.set(xlabel='ACT Participation')
ax.set(ylabel='Percentage')
ax.set_title('The Plot Value for ACT Participation 2017/2018')
bplot = sns.boxplot(
                 data=final[['act_participation_17','act_participation_18']], 
                 ax=ax)

In [ ]:
fig, ax=plt.subplots(figsize=(15,5))
ax.set(xlabel='ACT English')
ax.set(ylabel='ACT Score')
ax.set_title('The Plot Value for ACT English 2017/2018')
bplot = sns.boxplot(
                 data=final[['act_english_17','act_english_18']], 
                 ax=ax)

In [ ]:
fig, ax=plt.subplots(figsize=(15,5))
ax.set(xlabel='ACT Math')
ax.set(ylabel='ACT Score')
ax.set_title('The Plot Value for ACT Math 2017/2018')
bplot = sns.boxplot(
                 data=final[['act_math_17','act_math_18']], 
                 ax=ax)

In [ ]:
fig, ax=plt.subplots(figsize=(15,5))
ax.set(xlabel='ACT Reading')
ax.set(ylabel='ACT Score')
ax.set_title('The Plot Value for ACT Reading 2017/2018')
bplot = sns.boxplot(
                 data=final[['act_reading_17','act_reading_18']], 
                 ax=ax)

In [ ]:
fig, ax=plt.subplots(figsize=(15,5))
ax.set(xlabel='ACT Science')
ax.set(ylabel='ACT Score')
ax.set_title('The Plot Value for ACT Science 2017/2018')
bplot = sns.boxplot(
                 data=final[['act_science_17','act_science_18']], 
                 ax=ax)

In [ ]:
fig, ax=plt.subplots(figsize=(15,5))
ax.set(xlabel='ACT Composite')
ax.set(ylabel='ACT Score')
ax.set_title('The Plot Value for ACT Composite 2017/2018')
bplot = sns.boxplot(
                 data=final[['act_composite_17','act_composite_18']], 
                 ax=ax)

#### Feel free to do additional plots below
*(do research and choose your own chart types & variables)*

Are there any additional trends or relationships you haven't explored? Was there something interesting you saw that you'd like to dive further into? It's likely that there are a few more plots you might want to generate to support your narrative and recommendations that you are building toward. **As always, make sure you're interpreting your plots as you go**.

There seems to be a NEGATIVE correlation between Participation and Composite Score for ACT
Below is a scatter plot of ACT Participation vs.Composite for 2017, showing a decrease in Particiaption, but with an increase in Composite Score.

In [ ]:
x = final.act_composite_17
y = final.act_participation_17

# Plot
plt.scatter(x, y)
plt.title('ACT Participation vs.Composite for 2017')
plt.xlabel('ACT Composite Score for 2017')
plt.ylabel('ACT Participation for 2017')
plt.show()

Similarly,the same NEGATIVE correlation can be seen for 2018. 

In [ ]:
x = final.act_composite_18
y = final.act_participation_18

# Plot
plt.scatter(x, y)
plt.title('ACT Participation vs.Composite for 2018')
plt.xlabel('ACT Composite Score for 2018')
plt.ylabel('ACT Participation for 2018')
plt.show()

#### (Optional): Using Tableau, create a choropleth map for each variable using a map of the US. 

Save this plot as an image file in an images directory, provide a relative path, and insert the image into notebook in markdown.

## Descriptive and Inferential Statistics

#### Summarizing Distributions

Above, we used pandas `describe` to provide quick summary statistics of our numeric columns. We also demonstrated many visual relationships.

As data scientists, having a complete understanding of data is imperative prior to modeling.

While we will continue to build our analytic tools, we know that measures of *central tendency*, *spread*, and *shape/skewness* provide a quick summary of distributions.

For each variable in your data, summarize the underlying distributions (in words & statistics)
 - Be thorough in your verbal description of these distributions.
 - Be sure to back up these summaries with statistics.

Answers:

#### We generally assuming that data we sample from a population will be normally distributed. Do we observe this trend?

Answer: the histograms indicate that the sample data is NOT normally distributed. 

Does This Assumption Hold for:
    - Math
    - Reading
    - Rates
Explain your answers for each distribution and how you think this will affect estimates made from these data.

Answer: the histogram for the subjects and rates do NOT show normal distribution.

#### Estimate Limits of Data

Suppose we only seek to understand the relationship between SAT and ACT participation rates in 2017. 

##### Does it make sense to conduct statistical inference given these data specifically? 

Why or why not?

*(think about granularity, aggregation, the relationships between populations size & rates...consider the actually populations these data describe in answering this question)*

Answer: Data on SAT and ACT 2017 alone is enough to provide statistical inference as it provides descriptive statistical information. However, there is not enough data to provide insight on what is the underlying reason for the distribution and how it can be improved. 

##### Is it appropriate to compare *these* specific SAT and ACT math scores? 

Why or why not?

Answer:the comparison of the math scores for SAT and ACT is NOT appropriate to be compared as they do not have similar rates.  

#### Statistical Evaluation of Distributions 

**If you feel it's appropriate**, using methods we discussed in class, run hypothesis tests to compare variables of interest in our dataset. 

In [ ]:
# Code: 

## Outside Research

Based upon your observations, choose **three** states that demonstrate interesting trends in their SAT and/or ACT participation rates. Spend some time doing outside research on state policies that might influence these rates, and summarize your findings below. **Feel free to go back and create new plots that highlight these states of interest**. If you bring in any outside tables or charts, make sure you are explicit about having borrowed them. If you quote any text, make sure that it renders as being quoted. (Make sure that you cite your sources -- check with you local instructor for citation preferences).

## Conclusions and Recommendations

Based on your exploration of the data, what are you key takeaways and recommendations? Choose one state with a lower participation rate and provide a suggestion for how the College Board might increase participation amongst graduating seniors in this state. Are there additional data you desire that would better inform your investigations?

1.	From the data provided, we can compare the test results from year to year. This will indicate the year-to-year result at the state level. However, cross comparison between SAT and ACT is not advisable as they are NOT of the same scoring system nor are their contents the same.

2.  Further data is required to understand the underlying reasons for the test results. (ie; ethnicity, prevailing student challenges specific to each state, etc) . only when the underlying reasons are clear can proper recommendations be made for future improvements.